modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+      another List.
+
+      :arg other: A List containing elements of the same type as those
+        contained in this List.
+    */
+    proc extend(other: List(?t)) where t == eltType {
+      _enter();
+      _extendGeneric(other);
+      _leave();
+    }
+
+    /*
+      Extend this List by appending a copy of each element contained in an
+      array.
+
+      :arg other: An array containing elements of the same type as those
+        contained in this List.
+    */
+    proc extend(other: [?d] ?t) where t == eltType {
+      _enter();
+      _extendGeneric(other);
+      _leave();
+    }
+
+    /*
+      Insert an element at a given position in this List, shifting all elements
+      following that index one to the right.
+
+      .. warning::
+      
+        Inserting an element into this List may invalidate existing references
+        to the elements contained in this List.
+
+      :arg i: The index of the element at which to insert.
+      :arg x: The element to insert.
+
+      :throws IllegalArgumentError: If the given index is out of bounds.
+    */
+    proc insert(i: int, pragma "no auto destroy" in x: eltType) throws {
+      _enter();
+      try _boundsCheckLeaveOnThrow(i);
+      _expand(i);
+      ref src = x;
+      ref dst = _getRef(i);
+      _move(src, dst);
+      _size += 1;
+      _leave();
+    }
+
+    /*
+      Remove the first item from this List whose value is equal to x, shifting
+      all elements following the removed item one to the left.
+
+      .. warning::
+
+        Removing an element from this List may invalidate existing references
+        to the elements contained in this List.
+
+      :arg x: The element to remove.
+
+      :throws IllegalArgumentError: If there is no such item.
+    */
+    proc remove(x: eltType) throws {
+      _enter();
+
+      for i in 1.._size do {
+          ref item = _getRef(i);
+          if x == item then {
+            _destroy(item);
+            _collapse(i);
+            _size -= 1;
+            _leave();
+            return;
+          }
+      }
+
+      _leave();
+
+      const msg = "No such element in List: " + x:string;
+      throw new owned
+        IllegalArgumentError(msg);
+    }
+
+    /*
+      Remove the item at the given position in this List, and return it. If no
+      index is specified, remove and return the last item in this List.
+
+      .. warning::
+
+        Popping an element from this List will invalidate any reference to
+        the element taken while it was contained in this List.
+
+      :arg i: The index of the element to remove. Defaults to the last item
+        in this List.
+
+      :return: The item removed.
+
+      :throws IllegalArgumentError: If the given index is out of bounds.
+    */
+    proc pop(i: int=size): eltType throws {
+      _enter();
+      try _boundsCheckLeaveOnThrow(i);
+      // TODO: Is this fine, or do I need to disable initializers here?
```

Comment:
I do think it would be good to move from _getRef(i) into result. If you do so, you do not need to destroy element i. If you don't move (like you have here), then you need to destroy element i. I'd opt for the move.